Add ADR008: Obtaining and Launching Apalache (#1106)
informalsystems/quint: doc/adr008-managing-apalache.md

# ADR008: Obtaining and Launching Apalache from Quint

| Revision | Date       | Author           |
| :------- | :--------- | :--------------- |
| 1        | 2023-11-13 | Thomas Pani      |

## 1. Summary

This RFC presents options on how to obtain and manage Apalache for use with
the Quint `verify` command.

## 2. Context

Quint modules can be verified through bounded model-checking by issuing the
`quint verify` command. This invokes [Apalache][] via it's [Shai][] RPC
interface.

There's three outstanding questions concerning this integration:

1. How is Shai managed (launched, stopped, ...)?
2. How is the Apalache distribution obtained?
3. How are version dependencies between Quint and Apalache managed?

> Obviously, a decision on each of these questions impacts the other ones. We use
> this document to survey options, and to list advantages and disadvantages.

Reviewer: ```suggestion
Obviously, a decision on each of these questions impacts the other ones. We use
this document to survey options, and to list advantages and disadvantages, and to
document the proposed solution.
```

Based on the final section.

##### Launching of a system service by the user

As a special case, the user may manually launch a system service (§3.1.1
above), e.g., via `systemctl` on Linux or `launchctl` on macOS. The pros/cons
in this case fall at the intersection of both options.

#### 3.1.3 Managed by Quint

This option launches Shai from the Quint binary (or the VSCode extension).

This is similar to how bazel [runs its build server](https://bazel.build/run/client-server),
and to how the Metals VSCode plugin [manages the metals server](https://github.com/scalameta/metals-vscode/blob/edc446cc58dc809f3ed7cfde15b1e7fd7b5be115/packages/metals-vscode/src/extension.ts#L175C10-L175C30).

##### On-demand vs long-lived

As a sub-option, we would have to decide if Shai is launched on-demand, for each
verification invocation, or long-lived for as long as the VSCode extension is
active. bazel and vscode-metals employ a hybrid approach, where a long-lived
server is launched on-demand and torn down after an inactivity timeout. This
decision should take into consideration the memory-overhead of a long-lived
server vs. JVM startup time for an on-demand approach; at the time of writing,

> a freshly launched `apalache-mc server` takes ~125MB of memory, while JVM
> startup time is negligable (< 2sec).

Author: Then let's just record the facts:

```suggestion
a freshly launched `apalache-mc server` takes ~125MB of memory, and JVM
startup time is below 2 seconds.
```

2. Invoking an ecosystem package manager from Quint (see §3.2.2)
3. Pull the Apalache distribution from Github (see §3.2.3)

##### 1. Invoking the system-level package manager from Quint

Invoking a system-level package manager from Quint seems infeasible; we would
have to include logic for various platforms within Quint.

##### 2. Invoking an ecosystem package manager from Quint

Here, we would invoke `npm` or `coursier` from Quint.

Pros:

- Completely transparent to the user.
- Allows managing the required Apalache version from within Quint.
- No system-level dependency like above.

Cons:

- Assumes that a package manager (npm, coursier, ...) and the JRE are installed on the local system.
  - To work around this, we could distribute the package manager with Quint
    (for example, `vscode-metals` includes a `coursier` binary in its
    distribution).
- Spawning a package manager increases complexity by running a process whose
  behavior is outside of our control.

##### 3. Pulling the Apalache distribution from Github

We either hardcode a fixed Apalache release (and its artifact URL) in Quint, or
use [Github's REST API endpoint][] and/or [`octokit/request.js`][] to look up

> a matching release and download the JAR distribution directly.

Reviewer: Three additional options have occurred to me: 

1. We could maintain list of supported versions in a file somewhere (e.g., in the quint repo) and just request the file content, without having to hit the GH API. We could semi-automate updates to this page.
2. We could scrape the releases from https://github.com/informalsystems/quint/releases, again only requiring content of a served page, rather then going through the API.
3. We could require users to authenticate with github in order to use the apalache managed version. This would remove the rate-limiting issue, and we could also authenticate in this way in our CI.

Note that part of the problem in the previous implementation is that, iirc, every invocation of `verify` was hitting the GH API, meaning you could not invoke the subcommand more than 60 times an hour from the same IP address. If we only checked for updated versions periodically -- e.g., once per week, or when a user runs an `update` subcommand -- we would not hit the exact same problem (but other would persist).

Author: Interesting options! Should we record these in a separate issue? I'm a bit concerned that this PR is becoming the record of decisions, rather than the ADR itself 😅

Author: split off into #1248

Reviewer: Thank you for opening the issue!

Pros:

- Completely transparent to the user.
- Allows managing the required Apalache version from within Quint.
- No package manager dependency like above.

Cons:

- Assumes that the JRE is installed on the local system.

*Note: We did a prototype implementation querying the GitHub REST API in [#1115](https://github.com/informalsystems/quint/issues/1115).

> However, we observed CI issues due to the Github API's rate limiting as described in [#1124](https://github.com/informalsystems/quint/issues/1124). In pratice, the same issues can affect users (e.g., behind a shared IP) and may segnificantly impact UX of the `verify` command. As a countermeasure, we reverted to a hardcoded Quint version in [`4ceb7d8`](https://github.com/informalsystems/quint/commit/4ceb7d8be824ddc0a2c2a14e105baff446f71e72).*

Reviewer 2: ```suggestion
However, we observed CI issues due to the Github API's rate limiting as described in [#1124](https://github.com/informalsystems/quint/issues/1124). In pratice, the same issues can affect users (e.g., behind a shared IP) and may segnificantly impact UX of the `verify` command. As a countermeasure, we reverted to a hardcoded Apalache version in [`4ceb7d8`](https://github.com/informalsystems/quint/commit/4ceb7d8be824ddc0a2c2a14e105baff446f71e72).*
```
